pipeline-validation/custom-pipeline-validate-orca/src/main/java/com/opsmx/plugin/pipeline/validate/runtime/PipelineValidateAgentPreprocessor.java

```diff
@@ -0,0 +1,114 @@
+package com.opsmx.plugin.pipeline.validate.runtime;
+
+import com.google.gson.Gson;
+import com.netflix.spinnaker.kork.plugins.api.internal.SpinnakerExtensionPoint;
+import com.netflix.spinnaker.kork.web.exceptions.ValidationException;
+import com.netflix.spinnaker.orca.api.pipeline.ExecutionPreprocessor;
+
+
+import javax.annotation.Nonnull;
+
+import org.pf4j.Extension;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+@Extension
+@Component
+@ComponentScan("com.opsmx.plugin.pipeline.runtime")
+public class PipelineValidateAgentPreprocessor implements ExecutionPreprocessor, SpinnakerExtensionPoint {
+
+	private final Logger logger = LoggerFactory.getLogger(PipelineValidateAgentPreprocessor.class);
+
+	private final Gson gson = new Gson();
+
+	@Value("${pipeline.validate.enabled:false}")
+	private boolean isPipelineValidateEnabled;
+
+	@Override
+	public boolean supports(@Nonnull Map<String, Object> pipeline, @Nonnull Type type){
+		logger.debug("ExecutionPreprocessor Type :{}",type);
+		if(type.equals(Type.PIPELINE)){
+			return true;
+		}
+		return false;
+	}
+	@Override
+	public Map<String, Object> process(@Nonnull Map<String, Object> pipeline){
+		logger.debug("Start of the Pipeline Validation");
+		logger.debug("input Pipeline :{}", pipeline);
+		if (!isPipelineValidateEnabled) {
+			logger.debug("Pipeline validation not enabled, returning");
+			logger.debug("End of the Pipeline Validation");
+			return pipeline;
+		}
+		try {
+           List stages = (ArrayList)pipeline.get("stages");
+		   List<String> refIds = new ArrayList<String>();
+			List<String> allRequisiteStageRefIds = new ArrayList<String>();
+		   for(Object stageObject : stages){
+			   Map<String, Object> stage = (HashMap) stageObject;
+			   refIds.add(stage.get("refId").toString());
+			   List<String> requisiteStageRefIds = (ArrayList)stage.get("requisiteStageRefIds");
+			   allRequisiteStageRefIds.addAll(requisiteStageRefIds);
+		   }
+           validateRefIds(refIds);
+		   validateRequisiteStageRefIds(refIds,allRequisiteStageRefIds);
+		   validateRefIdAndRequisiteStageRefIds(stages);
+		} catch (Exception e) {
+			e.printStackTrace();
+			//logger.error("Exception occured : {}", e);
+			logger.debug("Some thing wrong While validate pipeline input: {}", pipeline);
```

Comment:
shoujldn't this be an error log ? somthing like "Failed to valudate pipeline...."